lib/streamlit/runtime/fragment.py

```diff
@@ -265,8 +265,7 @@ def wrapped_fragment():
                 ctx.current_fragment_id = prev_fragment_id
                 ctx.current_fragment_delta_path = []
 
-        if not ctx.fragment_storage.contains(fragment_id):
```

Comment:
I have removed the caching here, because it can lead to issues in case of an inner fragment accesses global variables, for example like this:

```python
@st.fragment
def outer_fragment2():
    c = st.container(border=True)
    c.button("rerun outer fragment2")

    @st.fragment
    def inner_fragment2():
        with st.container(border=True):
            st.write(c._active_dg._cursor.delta_path)
            c.write(f"inner fragment2: {uuid4()}")
            st.write("foo")
            st.button("rerun inner fragment2")

    inner_fragment2()
```

With the caching, `inner_fragment2` has an old reference to `c` and the delta_path it tries to write to is incorrect after clearing the container (by re-running the outer fragment) - hence we see the index-error popup issue in the browser.
I believe this might already be an issue today, but we don't see it that often because the inner elements are stale.